scripts/evaluate_heartbeat.py

```diff
@@ -67,13 +68,69 @@ def get_operator(staker_address: str, taco_application: ContractContainer) -> st
         return "Unknown"
 
 
-def get_ordinal_suffix(n: int) -> str:
-    """Return the ordinal suffix for a number (1st, 2nd, 3rd, 4th, etc.)."""
-    if 10 <= n % 100 <= 20:
-        suffix = "th"
+def get_node_version(staker_address: str, network_data: Dict[str, Any]) -> str:
+    # if this node is the one that provided the network data, return it directly
+    if network_data.get("staker_address") == staker_address:
+        return network_data.get("version", UNREACHABLE)
     else:
-        suffix = {1: "st", 2: "nd", 3: "rd"}.get(n % 10, "th")
-    return suffix
+        nodes = network_data["known_nodes"]
+        for node in nodes:
+            if node.get("staker_address") == staker_address:
+                rest_url = node["rest_url"]
+                try:
+                    node_status = requests.get(
+                        f"https://{rest_url}/status/",
+                        params={"json": "true"},
+                        verify=False,
+                        timeout=30,
+                    )
+
+                    # check for HTTP errors (4xx and 5xx)
+                    node_status.raise_for_status()
+
+                    return node_status.json().get("version", UNREACHABLE)
+                except (
+                    requests.ConnectionError,
+                    requests.exceptions.ReadTimeout,
+                    requests.HTTPError,
+                ):
+                    return UNREACHABLE
+
+        return UNREACHABLE
+
+
+def get_valid_versions() -> List[Version]:
+    """Fetches valid versions considering the update grace period."""
+    releases_url = "https://api.github.com/repos/nucypher/nucypher/releases"
+    releases_response = requests.get(releases_url)
+    releases_response.raise_for_status()
+    releases_response = releases_response.json()
+
+    valid_versions: List[Version] = []
+
+    # GitHub seems to return releases in published_at-descending-order
+    for i, release in enumerate(releases_response):
+        version_str = release.get("tag_name")
+        try:
+            version = Version(version_str)
+        except InvalidVersion:
+            click.secho(
+                f"⚠️  Failed to parse version from release {version_str}"
+                + " Not semantic versioning format?",
+                fg="red",
+            )
+            return []
+
+        if i == 0:
+            valid_versions.append(version)
+        else:
+            previous_release_date = datetime.fromisoformat(releases_response[i - 1]["published_at"])
+            deadline = previous_release_date + NODE_UPDATE_GRACE_PERIOD
+            # filter the versions that are considered valid to run
+            if deadline > datetime.now(tz=timezone.utc):
+                valid_versions.append(version)
```

Comment:
I think you can also be more efficient here and break out of the loop once a non-valid version is encountered. If the release response is in descending order, then all other previous versions would also be invalid.

So something like:


```suggestion
            if deadline > datetime.now(tz=timezone.utc):
                valid_versions.append(version)
            else:
                # invalid, so all prior versions are also invalid
                break
```